pkg/cmd/project/link/link.go

```diff
@@ -0,0 +1,194 @@
+	} `graphql:"linkProjectV2ToRepository(input:$input)"`
+}
+
+type linkProjectToTeamMutation struct {
+	LinkProjectV2ToTeam struct {
+		Team queries.Team `graphql:"team"`
+	} `graphql:"linkProjectV2ToTeam(input:$input)"`
+}
+
+func NewCmdLink(f *cmdutil.Factory, runF func(config linkConfig) error) *cobra.Command {
+	opts := linkOpts{}
+	linkCmd := &cobra.Command{
+		Short: "Link a project to a repository or a team",
+		Use:   "link [<number>] [flag]",
+		Example: heredoc.Doc(`
+			# link monalisa's project 1 to her repository "my_repo"
+			gh project link 1 --owner monalisa --repo my_repo
+
+			# link monalisa's organization's project 1 to her team "my_team"
+			gh project link 1 --owner my_organization --team my_team
+		`),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			client, err := client.New(f)
+			if err != nil {
+				return err
+			}
+
+			if len(args) == 1 {
+				num, err := strconv.ParseInt(args[0], 10, 32)
+				if err != nil {
+					return cmdutil.FlagErrorf("invalid number: %v", args[0])
+				}
+				opts.number = int32(num)
+			}
+
+			config := linkConfig{
+				httpClient: f.HttpClient,
+				client:     client,
+				opts:       opts,
+				io:         f.IOStreams,
+			}
+
+			if config.opts.repo != "" && config.opts.team != "" {
+				return fmt.Errorf("specify only one of `--repo` or `--team`")
+			} else if config.opts.repo == "" && config.opts.team == "" {
+				return fmt.Errorf("specify either `--repo` or `--team`")
+			}
+
+			// allow testing of the command without actually running it
+			if runF != nil {
+				return runF(config)
+			}
+			return runLink(config)
+		},
+	}
+
+	linkCmd.Flags().StringVar(&opts.owner, "owner", "", "Login of the owner. Use \"@me\" for the current user.")
+	linkCmd.Flags().StringVarP(&opts.repo, "repo", "R", "", "The repository to be linked to this project")
+	linkCmd.Flags().StringVarP(&opts.team, "team", "T", "", "The team to be linked to this project")
+	cmdutil.AddFormatFlags(linkCmd, &opts.exporter)
+
+	return linkCmd
+}
+
+func runLink(config linkConfig) error {
+	canPrompt := config.io.CanPrompt()
+	owner, err := config.client.NewOwner(canPrompt, config.opts.owner)
+	if err != nil {
+		return err
+	}
+
+	project, err := config.client.NewProject(canPrompt, owner, config.opts.number, false)
+	if err != nil {
+		return err
+	}
+	config.opts.projectID = project.ID
+
+	httpClient, err := config.httpClient()
+	if err != nil {
+		return err
+	}
+	c := api.NewClientFromHTTP(httpClient)
+
+	if config.opts.repo != "" {
+		return linkRepo(c, owner, config)
+	} else if config.opts.team != "" {
+		return linkTeam(c, owner, config)
+	}
+	return nil
+}
+
+func linkRepo(c *api.Client, owner *queries.Owner, config linkConfig) error {
+	repo, err := api.GitHubRepo(c, ghrepo.New(owner.Login, config.opts.repo))
+	if err != nil {
+		return err
+	}
+	config.opts.repoID = repo.ID
+
+	query, variable := linkRepoArgs(config)
+	err = config.client.Mutate("LinkProjectV2ToRepository", query, variable)
```

Comment:
How would you feel about creating a `LinkProjectToRepository` on to the `client`? Looking at `Mutate` it looks like it leaks the GQL implementation to the caller (the doc indicates some tech debt here and I agree).

The same for `LinkProjectToTeam` and the `Unlink` variants.

Reply:
@williammartin I moved the functions to the `client`, take a look at the latest changes 🚀